File: 백준/Gold/15789. CTP 왕국은 한솔 왕국을 이길 수 있을까？/CTP 왕국은 한솔 왕국을 이길 수 있을까？.js
const { readFileSync } = require('fs');

const input = readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const edges = input.slice(1, m + 1).map((row) => row.split(' ').map(Number));
const [c, h, k] = input[m + 1].split(' ').map(Number);

// union-find
const getParent = (parent, x) => {
  if (parent[x] === x) {
    return x;
  }

  parent[x] = getParent(parent, parent[x]);

  return parent[x];
};

const unionParent = (parent, a, b) => {
  a = getParent(parent, a);
  b = getParent(parent, b);

  if (a < b) {
    parent[b] = a;
  } else {
    parent[a] = b;
  }
};

const solution = (n, m, edges, c, h, k) => {
  const parents = Array.from({ length: n + 1 }, (_, i) => i);

  for (const [a, b] of edges) {
    unionParent(parents, a, b);
  }

  const cParent = getParent(parents, c);
  const hParent = getParent(parents, h);

  const counter = {};

  for (let i = 1; i <= n; i++) {
    const parent = getParent(parents, i);

    if (!counter[parent]) {
      counter[parent] = 0;
    }

    counter[parent]++;
  }

  const sorted = Object.entries(counter).sort((a, b) => b[1] - a[1]);

  let left = k;
  let power = counter[cParent];

  for (const [key, value] of sorted) {
    // parseInt 주의
    if (parseInt(key) === cParent || parseInt(key) === hParent) {
      continue;
    }

    if (left === 0) {
      break;
    }

    left -= 1;
    power += value;
  }

  return power;
};

console.log(solution(n, m, edges, c, h, k));

File: 백준/Gold/15789. CTP 왕국은 한솔 왕국을 이길 수 있을까？/CTP 왕국은 한솔 왕국을 이길 수 있을까？.py
import sys
from collections import Counter
input = lambda: sys.stdin.readline().rstrip()
miis = lambda: map(int, input().split())
sys.setrecursionlimit(10**5)

#########################################

def get_parent(x):
    if parent[x] == x:
        return x
    parent[x] = get_parent(parent[x]) # get_parent 거슬러 올라가면서 parent[x] 값도 갱신
    return parent[x]

def union_parent(a, b):
    a = get_parent(a)
    b = get_parent(b)
    
    if a < b: # 작은 쪽을 parent로 합의하자
        parent[b] = a
    else:
        parent[a] = b        

def same_parent(a, b):
    return get_parent(a) == get_parent(b)

#########################################

N, M = miis()
parent = [i for i in range(N+1)]
for _ in range(M):
    X, Y = miis()
    union_parent(X, Y)
for i in range(1, N+1):
    get_parent(i) # 이걸 해줘야되나??
C, H, K = miis()
counter = Counter(parent[1:])
l = sorted(counter.items(), key=lambda x:x[1], reverse=True)
l = [val for key, val in l if not (key == parent[C] or key == parent[H])]
print(counter[parent[C]] + sum(l[:K]))